syncer/interface.go

```diff
@@ -41,26 +41,9 @@ type fetchLogic interface {
 
 // fetcher is the interface to the low-level fetching.
 type fetcher interface {
-	GetMaliciousIDs(
-		context.Context,
-		[]p2p.Peer,
-		func([]byte, p2p.Peer),
-		func(error, p2p.Peer),
-	) error
-	GetLayerData(
-		context.Context,
-		[]p2p.Peer,
-		types.LayerID,
-		func([]byte, p2p.Peer),
-		func(error, p2p.Peer),
-	) error
-	GetLayerOpinions(
-		context.Context,
-		[]p2p.Peer,
-		types.LayerID,
-		func([]byte, p2p.Peer),
-		func(error, p2p.Peer),
-	) error
+	GetMaliciousIDs(context.Context, []p2p.Peer) <-chan fetch.Result
+	GetLayerData(context.Context, []p2p.Peer, types.LayerID) (<-chan fetch.Result, error)
+	GetLayerOpinions(context.Context, []p2p.Peer, types.LayerID) (<-chan fetch.Result, error)
```

Comment:
I think it would be easier to have these methods only take 1 peer as parameter and return the `data` of the response directly instead of a `fetch.Result`. This completely gets rid of `poll` and makes the caller side also simpler. As an example `PollLayerOpinions` becomes:

```go
func (d *DataFetch) PollLayerOpinions(
	ctx context.Context,
	lid types.LayerID,
	needCert bool,
	peers []p2p.Peer,
) ([]*fetch.LayerOpinion, []*types.Certificate, error) {
	logger := d.logger.WithContext(ctx).WithFields(lid)
	opinions := make([]*fetch.LayerOpinion, 0, len(peers))
	success := false
	var combinedErr error
	for _, peer := range peers {
		resp, err := d.fetcher.GetLayerOpinion(ctx, peer, lid)
		if err != nil {
			opnsPeerError.Inc()
			logger.With().Debug("received peer error for layer opinions", log.Err(err))
			if !success {
				combinedErr = errors.Join(combinedErr, err)
			}
			continue
		}

		logger.Debug("received layer opinions from peer")
		var lo fetch.LayerOpinion
		if err := codec.Decode(resp, &lo); err != nil {
			logger.With().Debug("error decoding LayerOpinion", log.Err(err))
			if !success {
				combinedErr = errors.Join(combinedErr, err)
			}
			continue
		}
		lo.SetPeer(peer)
		opinions = append(opinions, &lo)
		success = true
		combinedErr = nil
	}

	if !needCert {
		return opinions, nil, combinedErr
	}

	certs := make([]*types.Certificate, 0, len(opinions))
	peerCerts := map[types.BlockID][]p2p.Peer{}
	for _, opinion := range opinions {
		if opinion.Certified == nil {
			continue
		}
		if _, ok := peerCerts[*opinion.Certified]; !ok {
			peerCerts[*opinion.Certified] = []p2p.Peer{}
		}
		peerCerts[*opinion.Certified] = append(peerCerts[*opinion.Certified], opinion.Peer())
		// note that we want to fetch block certificate for types.EmptyBlockID as well,
		// but we don't need to register hash for the actual block fetching
		if *opinion.Certified != types.EmptyBlockID {
			d.fetcher.RegisterPeerHashes(
				opinion.Peer(),
				[]types.Hash32{opinion.Certified.AsHash32()},
			)
		}
	}
	for bid, bidPeers := range peerCerts {
		cert, err := d.fetcher.GetCert(ctx, lid, bid, bidPeers)
		if err != nil {
			certPeerError.Inc()
			continue
		}
		certs = append(certs, cert)
	}
	return opinions, certs, combinedErr
}
```

It removes the need to process data from the channel. And now since all the control flow is in `PollLayerOpinions` we could re-add an `errgroup` to process all concurrently that we want to process concurrently. `opinions` and `peerCerts` could for instance become channels that are being read from while layer opinions are still being fetched from peers.

Reply:
The code from your comment is oversimplified as it makes the queries serially. In reality, you would need to add an errgroup and collect results on a channel. But it still could be simpler then the current approach, let me test it out :+1: .

Reply:
That's what I meant by

> we could re-add an errgroup to process all concurrently that we want to process concurrently. opinions and peerCerts could for instance become channels that are being read from while layer opinions are still being fetched from peers.

🙂

Reply:
@fasmat Updated as suggested.